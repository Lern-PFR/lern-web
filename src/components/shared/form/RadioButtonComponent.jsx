import styled from 'enzyme';
import {
	border,
	display,
	layout,
	shadow,
	color,
} from 'styled-system';
import PropTypes from 'prop-types';
import { disabled as disabledStyle } from 'theme/buttonStyles';
import { jasmine, tuna } from 'theme/colors';

const StyledRadioButtonComponent = styled('input')(
	border,
	display,
	layout,
	shadow,
	color,
	{
		...radioButtonStyle,
		appearance: 'none',
		MozAppearance: 'none',
		WebkitAppearance: 'none',
		position: 'relative',
		borderColor: (({ disabled }) => (disabled ? tuna.default : 'initial')),
		backgroundColor: (({ disabled }) => (disabled ? jasmine.darker1 : 'initial')),

		'&:focus': {
			outline: 'none',
		},

		'&::before': {
			position: 'absolute',
			top: 0,
			left: 0,
			content: '',
			fontSize: '12px',
			fontWeight: 700,
			color: 'transparent',
			display: 'inline-block',
			width: '100%',
			height: '100%',
			textAlign: 'center',
			borderColor: (({ disabled }) => (disabled ? tuna.default : 'initial')),
		},

		'&:checked:before': {
			content: '"\\2713"',
			color: 'white',
			backgroundColor: (({ disabled }) => (disabled ? jasmine.darker1 : 'black')),
			borderColor: (({ disabled }) => (disabled ? tuna.default : 'initial')),
		},
	},
);

/**
* @name CheckboxComponent
* @description A component used to display styled checkbox element.
*
* @param {bool}	[disabled]		: Whether the checkbox is disabled.
* @param {bool}	[checked]		: Whether the checkbox is checked.
* @param {string}	id				: The id of the checkbox.
*/
const RadioButtonComponent = ({ id, checked, disabled, ...otherProps }) => (
	<StyledRadioButtonComponent {...otherProps} {...(disabled ? disabledStyle : {})} type="radio" disabled={disabled} defaultChecked={checked} id={id} />
);

RadioButtonComponent.propTypes = {
	id: PropTypes.string.isRequired,
	disabled: PropTypes.bool,
	checked: PropTypes.bool,
};

RadioButtonComponent.defaultProps = {
	disabled: false,
	checked: false,
};

export default RadioButtonComponent;