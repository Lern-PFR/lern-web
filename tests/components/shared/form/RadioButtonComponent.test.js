import { shallow } from 'enzyme';
import RadioButtonComponent from 'components/shared/form/RadioButtonComponent';
import { act, cleanup, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

describe('RadioButtonComponent', () => {
	afterEach(() => {
		cleanup();
	});

	it('should match previous snapshot', () => {
		const sut = (<RadioButtonComponent id="cb" name="rg" />);
		const wrapper = shallow(sut);

		expect(wrapper).toMatchSnapshot();
	});

	it('should match previous snapshot for checked state', () => {
		const sut = (<RadioButtonComponent checked id="cb" name="rg" />);
		const wrapper = shallow(sut);

		expect(wrapper).toMatchSnapshot();
	});

	it('should match previous snapshot for disabled state', () => {
		const sut = (<RadioButtonComponent disabled id="cb" name="rg" />);
		const wrapper = shallow(sut);

		expect(wrapper).toMatchSnapshot();
	});

	it('should call the onChange method when clicking on the radio button', async () => {
		const mockedOnChange = jest.fn(() => {});
		const sut = (<RadioButtonComponent onChange={mockedOnChange} id="cb" name="rg" />);
		render(sut);

		await act(async () => {
			userEvent.click(screen.getByRole('radio'));
		});

		expect(mockedOnChange).toHaveBeenCalled();
	});

	it('should not call the onChange method when clicking on the disabled radio button', async () => {
		const mockedOnChange = jest.fn(() => {});
		const sut = (<RadioButtonComponent onChange={mockedOnChange} disabled id="cb" name="rg" />);
		render(sut);

		await act(async () => {
			userEvent.click(screen.getByRole('radio'));
		});

		expect(mockedOnChange).not.toHaveBeenCalled();
	});
});
