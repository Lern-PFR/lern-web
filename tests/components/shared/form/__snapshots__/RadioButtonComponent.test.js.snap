// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButtonComponent should match previous snapshot 1`] = `
<styled.input
  defaultChecked={false}
  disabled={false}
  id="cb"
  name="rg"
  type="radio"
/>
`;

exports[`RadioButtonComponent should match previous snapshot for checked state 1`] = `
<styled.input
  defaultChecked={true}
  disabled={false}
  id="cb"
  name="rg"
  type="radio"
/>
`;

exports[`RadioButtonComponent should match previous snapshot for disabled state 1`] = `
<styled.input
  bg="F6F6F4"
  color="#8C8C8C"
  defaultChecked={false}
  disabled={true}
  id="cb"
  name="rg"
  type="radio"
/>
`;
